lib/daru/category.rb

```diff
@@ -0,0 +1,883 @@
+    #   and values as values to these parameters
+    # @example
+    #   dv = Daru::Vector.new [:a, 1, :a, 1, :c], type: :category
+    #   dv.summary
+    #   # => #<Daru::Vector(6)>
+    #   #         size            5
+    #   #   categories            3
+    #   #     max_freq            2
+    #   # max_category            a
+    #   #     min_freq            1
+    #   # min_category            c
+    def summary
+      values = {
+        size: size,
+        categories: categories.size,
+        max_freq: @cat_hash.values.map(&:size).max,
+        max_category: @cat_hash.keys.max_by { |cat| @cat_hash[cat].size },
+        min_freq: @cat_hash.values.map(&:size).min,
+        min_category: @cat_hash.keys.min_by { |cat| @cat_hash[cat].size }
+      }
+
+      Daru::Vector.new values
+    end
+
+    alias_method :describe, :summary
+
+    # Raises TypeError since the vector to be converted is already of
+    #   type category
+    # @return [TypeError] Raises an exception
+    def to_category
+      raise TypeError, 'The vector is already of type category.'
+    end
+
+    # Converts a category type vector to non category type vector
+    # @return [Daru::Vector] non category type vector
+    def to_non_category
+      Daru::Vector.new to_a, name: name, index: index
+    end
+
+    # Sets index of the vector
+    # @param [Daru::Index, Daru::MultiIndex, Daru::CategoricalIndex, Array, Range]
+    #   idx new index to assign to vector
+    # @return [Daru::Index, Daru::CategoricalIndex, Daru::MultiIndex] the index assigned
+    # @example
+    #   dv = Daru::Vector.new [1, 2, 3], type: :category
+    #   dv.index = 'a'..'c'
+    #   dv
+    #   # => #<Daru::Vector(3)>
+    #   #   a   1
+    #   #   b   2
+    #   #   c   3
+    def index= idx
+      @index = coerce_index idx
+    end
+
+    private
+
+    def validate_categories input_categories
+      raise ArgumentError, 'Input categories and speculated categories mismatch' unless
+        (categories - input_categories).empty?
+    end
+
+    def add_extra_categories extra_categories
+      extra_categories.each { |cat| @cat_hash[cat] = [] }
+    end
+
+    def initialize_core_attributes data
+      # Create a hash to map each category to positional indexes
+      categories = data.each_with_index.group_by(&:first)
+      @cat_hash = categories.map { |cat, group| [cat, group.map(&:last)] }.to_h
+
+      # Map each category to a unique integer for effective storage in @array
+      map_cat_int = categories.keys.each_with_index.to_h
+
+      # To link every instance to its category,
+      # it stores integer for every instance representing its category
+      @array = map_cat_int.values_at(*data)
+
+      # Include plotting functionality
+      self.plotting_library = Daru.plotting_library
+    end
+
+    def category_from_position position
+      cat_from_int @array[position]
+    end
+
+    def assert_ordered operation
+      # Change ArgumentError to something more expressive
```

Comment:
It is ok to leave this kind of comments, but please add TODO or FIXME tag to it.